src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2491 @@
+    histogram = state->histogram;
+    histogram_total_tokens = state->histogram_total_tokens;
+  }
+
+  bin_choice = crypto_rand_int(histogram_total_tokens);
+
+  /* Skip all the initial zero bins */
+  while (!histogram[curr_bin]) {
+    curr_bin++;
+  }
+  curr_weight = histogram[curr_bin];
+
+  // TODO: This is not constant-time. Pretty sure we don't
+  // really need it to be, though.
+  while (curr_weight < bin_choice) {
+    curr_bin++;
+    tor_assert(curr_bin < state->histogram_len);
+    curr_weight += histogram[curr_bin];
+  }
+
+  tor_assert(curr_bin < state->histogram_len);
+  tor_assert(histogram[curr_bin] > 0);
+
+  // Store this index to remove the token upon callback.
+  if (state->token_removal != CIRCPAD_TOKEN_REMOVAL_NONE) {
+    mi->chosen_bin = curr_bin;
+  }
+
+  if (curr_bin == CIRCPAD_INFINITY_BIN(state)) {
+    if (state->token_removal != CIRCPAD_TOKEN_REMOVAL_NONE &&
+        mi->histogram[curr_bin] > 0) {
+      mi->histogram[curr_bin]--;
+    }
+
+    // Infinity: Don't send a padding packet. Wait for a real packet
+    // and then see if our bins are empty or what else we should do.
+    return CIRCPAD_DELAY_INFINITE;
+  }
+
+  tor_assert(curr_bin < CIRCPAD_INFINITY_BIN(state));
+
+  bin_start = circpad_histogram_bin_to_usec(mi, curr_bin);
+  bin_end = circpad_histogram_bin_to_usec(mi, curr_bin+1);
+
+  /* Truncate the high bin in case it's the infinity bin:
+   * Don't actually schedule an "infinite"-1 delay */
+  bin_end = MIN(bin_end, start_usec+state->range_usec);
+
+  // Sample uniformly between histogram[i] to histogram[i+1]-1,
+  // but no need to sample if they are the same timeval (aka bin 0 or bin 1).
+  if (bin_end <= bin_start+1)
+    return bin_start;
+  else
+    return (circpad_delay_t)crypto_rand_uint64_range(bin_start, bin_end);
+}
+
+/**
+ * Sample a value from the specified probability distribution.
+ *
+ * This performs inverse transform sampling
+ * (https://en.wikipedia.org/wiki/Inverse_transform_sampling).
+ *
+ * XXX: These formulas were taken verbatim. Need a floating wizard
+ * to check them for catastropic cancellation and other issues (teor?).
+ * Also: is 32bits of double from [0.0,1.0) enough?
+ */
+static double
+circpad_distribution_sample(circpad_distribution_t dist)
+{
+  double p = 0;
+
+  switch (dist.type) {
+    case CIRCPAD_DIST_NONE:
+      return 0;
+    case CIRCPAD_DIST_UNIFORM:
+      p = crypto_rand_double();
+      // param2 is upper bound, param1 is lower
+      /* The subtraction is exact as long as param2 and param1 are less than
+       * 2**53. The multiplication is accurate as long as (param2 - param1)
+       * is less than 2**52. (And when they are large, the low bits aren't
+       * important.) The result covers the full range of outputs, as long as
+       * p has a resolution of 1/2**32 or greater. */
+      p *= (dist.param2 - dist.param1);
+      p += dist.param1;
+      return p;
+    case CIRCPAD_DIST_LOGISTIC:
```

Comment:
Please see my commit e3ed32e9f6 for some tests that cover this function. The tests fail because some of these probability distribution functions are busted. In particular the LOGISTIC distribution returns elements out of range, and same for the GEOMETRIC distribution,  whereas the  pareto distribution seems to always return the same value?

Reply:
The prob distr functions were not busted. I was busted.
In any case, these tests are replaced by the new Riastradh code.